### PR TITLE
Add unique username generation and tweak username normalization

### PR DESCRIPTION
**Pull request by vdavid**
Based on [this Slack thread](https://sourcegraph.slack.com/archives/C07KZF47K/p1676567906311189), there seems to be a need for username normalization and unique username creation on the back end.
I set out to add one but realized there was already a function called [`NormalizeUsername`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/auth/auth.go?L74) which does a decent job and is well tested.

So I did three things:
- Modified `NormalizeUsername` to accept more strings (throw less frequently) in a backward-compatible manner.
- Added `AddRandomSuffix` to assist with generating random valid usernames.
- Added `IsValidUsername` to check if a username is valid against DB constraints.

This allows for a sequence of:
- `IsValidUsername` → `NormalizeUsername` if needed → check if username already exists in DB → `AddRandomSuffix` → (if you want to be extra safe, do the random suffix call in a loop until it's unique)

Note: originally, in the case of email addresses, I intended to just replace the `@` with an underscore or something, but the existing solution chopped it off, which, I think, is an equally good solution if not better (in standard corporate email catalogs, the username is usually before the `@`), so I kept it.

## Test plan

All the new additions are well tested

### CODE REVIEW

File: cmd/frontend/auth/auth.go [hunk 1]
```diff
@@ -61,24 +62,34 @@ func composeMiddleware(middlewares ...*Middleware) *Middleware {
 // username formatting rules (based on, but not identical to
 // https://help.github.com/enterprise/2.11/admin/guides/user-management/using-ldap/#username-considerations-with-ldap):
 //
-// - Any characters not in `[a-zA-Z0-9-._]` are replaced with `-`
+// - Any characters not in `[a-zA-Z0-9-.]` are replaced with `-`
```
Comment (eseliger) [inline cmd/frontend/auth/auth.go:65]: why do we now replace underscores, too?

File: cmd/frontend/auth/auth.go [hunk 2]
```diff
@@ -90,6 +101,34 @@ func NormalizeUsername(name string) (string, error) {
...
-	disallowedSymbols   = lazyregexp.New(`(^[\-\.])|(\.$)|([\-\.]{2,})`)
-	disallowedCharacter = lazyregexp.New(`[^\w\-\.]`)
+	disallowedCharacter       = lazyregexp.New(`[^\w\-\.]`)
+	consecutivePeriodsSlashes = lazyregexp.New(`[\-\.]{2,}`)
+	sequencesToTrim           = lazyregexp.New(`(^[\-\.])|(\.$)|`)
 )
+
+// AddRandomSuffix appends a random 5-character lowercase alphabetical suffix (like "-lbwwt")
+// to the username to avoid collisions. If the username already ends with a dash, it is not
+// added again.
```
Comment (eseliger) [inline cmd/frontend/auth/auth.go:111]: wonder if a numeric suffix might be slightly nicer, or a "add human readable word" or so approach, but not a blocker

File: cmd/frontend/auth/auth.go [hunk 3]
```diff
@@ -90,6 +101,34 @@ func NormalizeUsername(name string) (string, error) {
 }
 
 var (
-	disallowedSymbols   = lazyregexp.New(`(^[\-\.])|(\.$)|([\-\.]{2,})`)
-	disallowedCharacter = lazyregexp.New(`[^\w\-\.]`)
+	disallowedCharacter       = lazyregexp.New(`[^\w\-\.]`)
+	consecutivePeriodsSlashes = lazyregexp.New(`[\-\.]{2,}`)
```
Comment (sashaostrikov) [inline cmd/frontend/auth/auth.go:105]:
```suggestion
	consecutivePeriodsDashes = lazyregexp.New(`[\-\.]{2,}`)
```
Should it be like this?

File: cmd/frontend/auth/auth.go [hunk 4]
```diff
@@ -61,24 +62,34 @@ func composeMiddleware(middlewares ...*Middleware) *Middleware {
 // username formatting rules (based on, but not identical to
 // https://help.github.com/enterprise/2.11/admin/guides/user-management/using-ldap/#username-considerations-with-ldap):
```
Comment (eseliger) [inline cmd/frontend/auth/auth.go:63]: sad, this like seems to have broken :D was curious what they did.

**Review by eseliger — APPROVED** [review]

**Review by sashaostrikov — APPROVED** [review]
LGTM!
I'd add a changelog entry if we go for adding replacements for underscores.

File: cmd/frontend/auth/auth.go [hunk 5]
```diff
@@ -90,6 +101,34 @@ func NormalizeUsername(name string) (string, error) {
...
+	}
+	return username + "-" + string(b), nil
+}
+
+// Equivalent to `^\w(?:\w|[-.](?=\w))*-?$` which we have in the DB constraint, but without a lookahead
+var validUsername = lazyregexp.New(`^\w(?:(?:[\w.-]\w|\w)*-?|)$`)
+
+// IsValidUsername returns true if the username matches the constraints in the database.
+func IsValidUsername(name string) bool {
+	return validUsername.MatchString(name) && len(name) <= 255
```
[the remark this answers is not in the document]
Comment (eseliger) [reply cmd/frontend/auth/auth.go:132]: in the slack thread on this topic, one suggestion for email addresses as usernames was to do the following:

- Add the email address as a `user_external_accounts` entry that keeps track of the auth provider identity
- Create a normalized username that does fit in the Sourcegraph username constraints (David was talking about using the username of the email address, plus normalization rules in here like the rand suffix, so `erik@sourcegraph.com` -> `erik` or `erik-asdf` if `erik` is taken)

Would that help you?

File: cmd/frontend/auth/auth.go (hunk 2 above)
Comment (indradhanush) [reply cmd/frontend/auth/auth.go:111]: @eseliger The internet never forgets: https://web.archive.org/web/20180215000330/https://help.github.com/enterprise/2.11/admin/guides/user-management/using-ldap/ 😉

File: cmd/frontend/auth/auth.go (hunk 4 above)
Comment (indradhanush) [reply cmd/frontend/auth/auth.go:63]: @eseliger The internet never forgets: https://web.archive.org/web/20180215000330/https://help.github.com/enterprise/2.11/admin/guides/user-management/using-ldap/ 😉

File: cmd/frontend/auth/auth.go (hunk 2 above)
Comment (indradhanush) [reply cmd/frontend/auth/auth.go:111]: Agreed. I had the same thought. Numberical suffixes are nicer and a more obvious indication that there was a conflict which made it add a numerical suffix. A random 5 char suffix might look like a bot account.

+1 to numerical suffixes if we can do it, although then this is no longer doable without touching the DB.

File: cmd/frontend/auth/auth.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (indradhanush) [reply cmd/frontend/auth/auth.go:132]: Yeah, technically the max chars in an email seems to be 254 two more for the (`<` and `>`) so this is fine as it is. 👍

[the remark this answers is not in the document]
Comment (indradhanush) [reply cmd/frontend/auth/auth.go:132]: Oops, I did not refresh my page before commenting above.

[the remark this answers is not in the document]
Comment (indradhanush) [reply cmd/frontend/auth/auth.go:132]: > Create a normalized username that does fit in the Sourcegraph username constraints (David was talking about using the username of the email address, plus normalization rules in here like the rand suffix, so erik@sourcegraph.com -> erik or erik-asdf if erik is taken)
> 
> Would that help you?

Yes, I think that would make sense. I will create an issue to make sure I change the current behaviour to use email address and instead use a `normalizeEmailToUsername` equivalent before releasing this feature for beta / public use.

File: cmd/frontend/auth/auth.go (hunk 1 above)
Comment (vdavid) [reply cmd/frontend/auth/auth.go]: We don't. It was different in this comment and the [docs](https://docs.sourcegraph.com/admin/auth#username-normalization), and I fixed it at the wrong place because I misunderstood the sentence. 😬  Reverted this change and changed it in the docs instead. Thanks @eseliger!

File: cmd/frontend/auth/auth.go [hunk 6]
```diff
@@ -90,6 +101,34 @@ func NormalizeUsername(name string) (string, error) {
...
+	consecutivePeriodsSlashes = lazyregexp.New(`[\-\.]{2,}`)
+	sequencesToTrim           = lazyregexp.New(`(^[\-\.])|(\.$)|`)
 )
+
+// AddRandomSuffix appends a random 5-character lowercase alphabetical suffix (like "-lbwwt")
+// to the username to avoid collisions. If the username already ends with a dash, it is not
+// added again.
+func AddRandomSuffix(username string) (string, error) {
+	b := make([]byte, 5)
+	_, err := rand.Read(b)
```
Comment (mrnugget) [inline cmd/frontend/auth/auth.go:114]: Where is that seeded? In the tests you call `rand.Seed` explicitly. Does that also have to happen in this package?

File: cmd/frontend/auth/auth.go (hunk 2 above)
Comment (vdavid) [reply cmd/frontend/auth/auth.go:111]: My main concern is that (if by "numerical suffix" we mean an _increasing_ numerical suffix) then in the odd case of adding `N` usernames that resolve to the same normalized username, we need `O(N²)` trips to the DB. Might bee too theoretical a concern, but wanted to avoid it.
_Unless_ we write a binary search for the next username that doesn't exist ;) I'm sure that implementation exists in a few systems 😄 

As this is a decision that's easy to change later, we could go with the current solution for now.

File: cmd/frontend/auth/auth.go (hunk 3 above)
Comment (vdavid) [reply cmd/frontend/auth/auth.go]: No, sorry, the underscore change was a documentation mistake, I didn't actually change that behavior.
Thanks for the variable name correction, it should be "dashes", fixed it. 🙂

File: cmd/frontend/auth/auth.go (hunk 6 above)
Comment (vdavid) [reply cmd/frontend/auth/auth.go:114]: In the tests I want it to be predictable, hence I seed it there.
In the production code, I want it to be unpredictable, so I don't seed it.
(On the other hand, I don't want cryptographically secure random numbers, hence I just used `math/rand`, not `bcrypt`.)
So from a seeding POV, the code does what I intended. Am I missing something?

File: cmd/frontend/auth/auth.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (vdavid) [reply cmd/frontend/auth/auth.go:132]: @indradhanush we have a constraint in Postgres called `users_username_max_length` which is `char_length((username)::text) <= 255`. I matched this with the validation code.
I _think_ it won't mess up anything if you increase this constraint (if that turns out to be the right solution to your problem), and then you can also increase it here.
But doing the DB change is out of scope for this PR, so I'll leave it at `<=255` for now.

[the remark this answers is not in the document]
Comment (indradhanush) [reply cmd/frontend/auth/auth.go:132]: Yes, that makes sense. We should stick to 255. Sorry I forgot to mark this as resolved after replying to Erik.

File: cmd/frontend/auth/auth.go (hunk 6 above)
Comment (vdavid) [reply cmd/frontend/auth/auth.go:114]: I'll merge this to go forward, but happy to fix it in a follow-up if I messed up something around seeding that I didn't realize. At the point of merging it, this is unused code, so very low stakes.

Comment (mrnugget) [reply cmd/frontend/auth/auth.go:114]: Ah, got it. Sorry, misunderstood `Seed`. While looking it up I noticed that it's deprecated though: https://pkg.go.dev/math/rand#Seed

File: cmd/frontend/auth/auth.go [hunk 7]
```diff
@@ -59,26 +60,36 @@ func composeMiddleware(middlewares ...*Middleware) *Middleware {
 
 // NormalizeUsername normalizes a proposed username into a format that meets Sourcegraph's
 // username formatting rules (based on, but not identical to
-// https://help.github.com/enterprise/2.11/admin/guides/user-management/using-ldap/#username-considerations-with-ldap):
+// https://web.archive.org/web/20180215000330/https://help.github.com/enterprise/2.11/admin/guides/user-management/using-ldap):
```
Comment (unknwon) [inline cmd/frontend/auth/auth.go:63]: ❤️

**Review by unknwon — COMMENTED** [review]
Thank you!

File: cmd/frontend/auth/auth.go (hunk 6 above)
Comment (vdavid) [reply cmd/frontend/auth/auth.go:114]: I understand why Seed is deprecated and how this _could_, in theory, cause a flaky test in an edge case, but in reality, knowing the code, I know it won't ever be flaky, so leaving that deprecated reference in there for the sake of moving fast. Thanks for spotting that, though, I didn't notice it!
